Refactor `AuthorizationCheckBehavior`

ce-dmelnych opened the pull request:
> [!NOTE]  
> Don't be alarmed by the large number of changed lines (`+28xx` / `-12xx`)!
> GitHub's diff algorithm incorrectly shows `AuthorizationCheckBehavior` as deleted and recreated as a new file.  In reality, the class was moved to a sub-package (`~.identity.authorization`) and modified by extracting code into dedicated classes.

> [!TIP]
> Please review this PR commit by commit, GitHub displays the changes correctly when viewing individual commits, making it much easier to follow the actual refactoring steps.

## Description

This PR refactors the `AuthorizationCheckBehavior` class to reduce its complexity and prepare the codebase for upcoming authorization enhancements. The class had grown to 800+ lines with multiple inner classes, records, enums, and mixed responsibilities.

### What was done

#### Package structure & type extraction

1. **Created dedicated package:** `io.camunda.zeebe.engine.processing.identity.authorization`

2. **Extracted to `request/` subpackage:**
   - `AuthorizationRequest` — converted from mutable class to immutable record with Builder pattern
   - Removed `AuthorizationRequestMetadata` (merged into `AuthorizationRequest`)

3. **Extracted to `result/` subpackage:**
   - `AuthorizationResult` record
   - `AuthorizationRejection` record
     - `AuthorizationRejectionType` enum

4. **Extracted `ForbiddenException`** as standalone class

5. **Removed unused exceptions:** `NotFoundException`, `AuthorizationsNotFoundException`

#### Component Extraction

6. **Extracted to `resolver/` subpackage:**
   - `ClaimsExtractor` — utility for extracting username, clientId, groups from claims
   - `TenantResolver` — resolving authorized tenants for entities
   - `AuthorizationScopeResolver` — resolving authorization scopes (direct + inherited via roles/groups)

#### API Improvements

7.  **Replaced `authorizationScopes` with `resourceIds`** in `AuthorizationRequest`:
   - Changed `Set<AuthorizationScope>` to `Set<String>` in request
   - Moved wildcard scope handling to `AuthorizationCheckBehavior#isAuthorizedByScopes()`
   - Prepares for future `resourceProperties` field for property-based authorization

8. **Added `isInternalCommand()` to `TypedRecord` interface:**
   - Internal command detection now belongs with the command itself
   - Simplifies call sites from parameter passing to `command.isInternalCommand()`

### Why this refactoring was needed

This refactoring is a prerequisite for upcoming authorization enhancements:
- **Property-based authorization [#40114](https://github.com/camunda/camunda/issues/40114)**: Simplified `AuthorizationRequest` with `resourceIds` instead of `AuthorizationScope` makes adding `resourceProperties` straightforward.

Also, it provides:
- **Improved maintainability**: Smaller, focused components are easier to understand, test, and modify independently.
- **Better testability**: Extracted components with single responsibilities can be unit tested in isolation.

## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [ ] Enable backports when necessary (fex. [for bug fixes](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#backporting-changes) or [for CI changes](https://github.com/camunda/camunda/wiki/CI-&-Automation#when-to-backport-ci-changes)).

## Related issues

closes #41854

ce-dmelnych committed:
refactor: move AuthorizationCheckBehavior to dedicated package

Create new package for authorization components decomposition.

ce-dmelnych committed:
refactor: remove unused exception classes in AuthorizationCheckBehavior

Delete `NotFoundException` and `AuthorizationsNotFoundException` - no usages found.

ce-dmelnych committed:
refactor: extract authorization result types

Extract inner records and enums from AuthorizationCheckBehavior.

ce-dmelnych committed:
refactor: extract ForbiddenException class

Extract `ForbiddenException` from `AuthorizationCheckBehavior`.

ce-dmelnych committed:
refactor: extract request classes to `request` package

Extract request-related classes from `AuthorizationCheckBehavior` to dedicated package:
- AuthorizationRequestMetadata: immutable request data
- AuthorizationRequest: builder class

ce-dmelnych committed:
refactor: move `isInternalCommand` to TypedRecord interface

Move internal command detection logic from `AuthorizationCheckBehavior`
to `TypedRecord` interface where it logically belongs.

Benefits:
- `TypedRecord` has direct access to command metadata
- Simplifies call sites: `command.isInternalCommand()` vs
  `authCheckBehavior.isInternalCommand(command.hasRequestMetadata(),
  command.getBatchOperationReference())`
- Removes authorization-unrelated logic from `AuthorizationCheckBehavior`

ce-dmelnych committed:
refactor: convert AuthorizationRequest to immutable record with Builder

Transform `AuthorizationRequest` from a mutable builder-like class to an
immutable record with a proper `Builder` pattern.

Changes:
- Convert `AuthorizationRequest` class to record
- Add `AuthorizationRequest.Builder` for fluent construction
- Add `AuthorizationRequest.of(Consumer<Builder>)` factory method
- Add `isTriggeredByInternalCommand` field (computed from command)
- Add convenience methods: `newResource()`, `addAllResourceIds()`
- Adjust `AuthorizationCheckBehavior#isAuthorizedOrInternalCommand()` to perform check
  for internal command using `isTriggeredByInternalCommand` flag
- Remove `AuthorizationRequestMetadata` (merged into `AuthorizationRequest`)
- Remove deprecated `AuthorizationCheckBehavior#isAuthorized(AuthorizationRequest)` overload
- Update all call sites to use builder pattern

ce-dmelnych committed:
refactor: replace authorizationScopes with resourceIds in AuthRequest

The previous implementation required adding `AuthorizationScope.WILDCARD` to every
request to ensure wildcard permissions would match.
This seemed like a hacky workaround that placed part of scope-matching responsibility in the wrong place,
the `AuthorizationRequest` was responsible for knowing how matching works, rather than simply declaring what
resources it needs access to.

This refactoring replaces `Set<AuthorizationScope>` with `Set<String> resourceIds` in
`AuthorizationRequest` and moves the wildcard matching logic to a new
`AuthorizationCheckBehavior#isAuthorizedByScopes()` method.
This method is responsible for handling all cases:
- denying access when no scopes are granted
- allowing access when `WILDCARD` is granted
- requiring wildcard permission when no specific resourceIds are requested,
  and otherwise matching granted `resourceIds` against requested ones.

This change simplifies the `AuthorizationRequest` model and prepares for property-based
authorization. Adding `resourceProperties` in follow-up PRs will now be straightforward
since the request simply declares identifiers rather than scope objects.

ce-dmelnych committed:
refactor: extract ClaimsExtractor from AuthorizationCheckBehavior

Extract claims-related helper methods to dedicated utility class.
This improves testability and separates claims parsing from authorization logic.

ce-dmelnych committed:
refactor: extract TenantResolver from AuthorizationCheckBehavior

Extract tenant resolution logic into a dedicated `TenantResolver` class.

This separation makes the code more maintainable and easier to test,
while keeping authorization validation logic in `AuthorizationCheckBehavior`.

ce-dmelnych committed:
refactor: extract AuthorizationScopeResolver

Extract authorization scope resolution logic from `AuthorizationCheckBehavior` into a dedicated
`AuthorizationScopeResolver` class.

This separation makes the code more maintainable and easier to test,
while keeping authorization validation logic in `AuthorizationCheckBehavior`.

npepinpe commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/identity/authorization/request/AuthorizationRequest.java:
```diff
@@ -0,0 +1,142 @@
+    return new AuthorizationRequest.Builder();
+  }
+
+  public static AuthorizationRequest of(final Consumer<Builder> consumer) {
```
❌ I don't think this pattern brings us much, and it creates an unnecessary lambda without benefits (but the downsides of lambdas).

What's the benefit over just calling

```java
AuthorizationRequest
  .builder()
  .command(command)
  .resourceType(type)
  .build();
```

Unless I'm missing something, then please revert or inline these usages like this.

ce-dmelnych replied:
I agree and also had some doubts about this pattern. Initially, I was inspired by similar methods in [`FollowUpEventMetadata.java`](https://github.com/camunda/camunda/blob/aa1ef7ec426c30aa9a0f5e34753dfcfad5e439dc/zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/FollowUpEventMetadata.java#L30-L34), [`Authorization.java`](https://github.com/camunda/camunda/blob/aa1ef7ec426c30aa9a0f5e34753dfcfad5e439dc/security/security-core/src/main/java/io/camunda/security/auth/Authorization.java#L67-L69) and so on. 

The `.of(r -> r.xxx)` pattern looks nicely when we need to set 1-2 properties (e.g. [like here](https://github.com/camunda/camunda/blob/aa1ef7ec426c30aa9a0f5e34753dfcfad5e439dc/zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/batchoperation/BatchOperationFailProcessor.java#L96)) where it fits on a single line and is used inline (e.g., passed directly as a method parameter). In our case, `AuthorizationRequest` usually requires 3+ properties and is mostly assigned to a separate variable.  Additionally, the `spotless` plugin formats it in a strange way, putting the parameter name and `->` on separate lines, which doesn't look great.

So I agree that using the plain `...builder().xxx(). build()` pattern is better here.  Adjusted all places to use the builder.

npepinpe commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/identity/authorization/request/AuthorizationRequest.java:
```diff
@@ -0,0 +1,142 @@
+      final Map<String, Object> claims =
+          command != null ? command.getAuthorizations() : authorizationClaims;
+      final boolean isTenantOwnedResource = tenantId != null && !tenantId.isEmpty();
+      final boolean triggeredByInternalCommand = command != null && command.isInternalCommand();
```
❓ Does this mean if someone does not use the builder, the values may be incorrect? What would happen in this case? Should we be more defensive and either:

1. Not use a builder but do everything in the constructor, or
2. Assert the values in the constructor, or
3. Enforce usage of the builder by making the constructor private.

npepinpe reviewed (changes requested):
❓ Is it correct that you did not change the logic of the split authorization check behavior? e.g. `TenantResolver`, `ClaimsExtractor`, etc.? I did not thoroughly check that as I assumed you kept the business logic identical.

ce-dmelnych committed:
refactor: replace AuthorizationRequest.of() with builder pattern

Replace `AuthorizationRequest.of(r -> r.xxx())` pattern with the standard
builder pattern `AuthorizationRequest.builder().xxx().build()` for
improved readability in multi-line configurations.

Rationale:
While the functional-style `of(..)` pattern is concise for single-line
configurations, it becomes less convenient when constructing requests
with multiple properties across multiple lines. The standard builder
pattern provides better clarity and consistency with other builder
usages in the codebase.

No functional changes.

ce-dmelnych committed:
refactor: enforce builder pattern for AuthorizationRequest

Convert `AuthorizationRequest` from record to immutable class with private
constructor to prevent direct instantiation. This ensures all instances
are created via the `builder`, which properly computes derived properties
like `isTenantOwnedResource` and `isTriggeredByInternalCommand`.

ce-dmelnych commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/identity/authorization/result/AuthorizationRejection.java:
```diff
@@ -0,0 +1,40 @@
+/**
+ * Represents a typed authorization failure which occurred while evaluating an authorization check.
+ */
+public record AuthorizationRejection(
```
Nice, I like this sealed interface approach! Applied ✅.

ce-dmelnych committed:
refactor: use sealed interface for AuthorizationRejection

Replace record + enum pattern with sealed interface and inner records.
This approach encodes the rejection type directly in the type hierarchy.
